src/network/connection_handle.cpp

```diff
@@ -1,188 +1,257 @@
+      default:                              throw std::runtime_error("Undefined transition!");
+    }
+  }
+
+  /** Implement transition for ConnState::PROCESS. */
+  static ConnectionHandle::StateMachine::TransitionResult TransitionForProcess(Transition transition) {
+    switch (transition) {
+      case Transition::NEED_READ:           return {ConnState::READ, TryRead};
+      case Transition::NEED_READ_TIMEOUT:   return {ConnState::READ, WaitForReadWithTimeout};
+      case Transition::NEED_RESULT:         return {ConnState::PROCESS, WaitForTerrier};
+      case Transition::PROCEED:             return {ConnState::WRITE, TryWrite};
+      case Transition::TERMINATE:           return {ConnState::CLOSING, TryCloseConnection};
+      case Transition::WAKEUP:              return {ConnState::PROCESS, GetResult};
+      default:                              throw std::runtime_error("Undefined transition!");
+    }
+  }
+
+  /** Implement transition for ConnState::WRITE. */
+  static ConnectionHandle::StateMachine::TransitionResult TransitionForWrite(Transition transition) {
+    switch (transition) {
+        // Allegedly, NEED_READ happens during ssl-rehandshake with the client.
+      case Transition::NEED_READ:           return {ConnState::WRITE, WaitForRead};
+      case Transition::NEED_WRITE:          return {ConnState::WRITE, WaitForWrite};
+      case Transition::PROCEED:             return {ConnState::PROCESS, Process};
+      case Transition::TERMINATE:           return {ConnState::CLOSING, TryCloseConnection};
+      case Transition::WAKEUP:              return {ConnState::WRITE, TryWrite};
+      default:                              throw std::runtime_error("Undefined transition!");
+    }
+  }
+
+  /** Implement transition for ConnState::CLOSING. */
+  static ConnectionHandle::StateMachine::TransitionResult TransitionForClosing(Transition transition) {
+    switch (transition) {
+      case Transition::NEED_READ:           return {ConnState::WRITE, WaitForRead};
+      case Transition::NEED_WRITE:          return {ConnState::WRITE, WaitForWrite};
+      case Transition::TERMINATE:           return {ConnState::CLOSING, TryCloseConnection};
+      case Transition::WAKEUP:              return {ConnState::CLOSING, TryCloseConnection};
+      default:                              throw std::runtime_error("Undefined transition!");
+    }
+  }
+  // clang-format on
+
+ private:
+  /** Define a function (ManagedPointer<ConnectionHandle> -> Transition) that calls the underlying handle's function. */
+#define DEF_HANDLE_WRAPPER_FN(function_name)                                               \
+  static Transition function_name(const common::ManagedPointer<ConnectionHandle> handle) { \
+    return handle->function_name();                                                        \
+  }
+
+  DEF_HANDLE_WRAPPER_FN(GetResult);
+  DEF_HANDLE_WRAPPER_FN(Process);
+  DEF_HANDLE_WRAPPER_FN(TryRead);
+  DEF_HANDLE_WRAPPER_FN(TryWrite);
```

Comment:
Is `TryWrite` ever even used?

Reply:
In Process::PROCESS and Write::WAKEUP, the function is called.
I don't know off the top of my head if those two states occur, but it doesn't seem implausible.
A breakpoint on TryWrite does get hit.